### DISCUSSION TITLE
Why ServerStop takes 10 seconds?

### DISCUSSION THREAD

**Discussion by 28itay** [pymodbus-dev/pymodbus#1719] [Help] [answered]
I know why, it has 10 seconds sleep in it, but why does it have 10 seconds sleep?

**Comment by janiversen** [comment, marked as answer]
Because that is a synchronous call, and the asynchronous part needs to time be scheduled and close down. Especially on windows this takes forever.

Use async if you want something faster.

**Reply by 28itay** [reply]
Thanks for the really quick answer, I have tried using async, but I can't seem to make it work, the server
I'm using `StartTcpServer` in a thread and `asyncio.run(ServerAsyncStop())` to stop it, but the server never stop running...

**Reply by janiversen** [reply]
That is not how you use async !

StartAsyncTcpServer works with ServerStop
StartTcpServer works with ServerAsyncStop

Actually quite logical or not ?

**Reply by 28itay** [reply]
`StartTcpServer` just calls `asyncio.run(StartAsyncTcpServer())`, I just don't get it...

Still don't get why it takes 10 seconds?, why it's "hard" coded and has nothing to do with how long it actually take to shutdown, I see the server get terminated in less than a second (or when serve_forever end doesn't mean the server is terminated?)

**Comment by janiversen** [comment]
Might be you do not get it, but just because StartTcpServer calls asyncio.run does not mean you can do it multiple times (as you tried), that is simply not how asyncio works, please read up on asyncio before posting.

Don't use startServer is you don't like it, that is a convenience function on top of the class. You can simply instantiate the ModbusTcpServer class, and manage it as you like, with the delays you want. Or make a pull request with a better solution, something that actually tests when the asyncio loop is actually closed.

Have you tried all the python version on all the platforms we support, when you claim it takes less than a second. We have and we actually do every time we run CI, so there is a reason for the delay.

**Reply by 28itay** [reply]
When and how did I do it multiple times????

I used your package in python 3.5, and it worked, I now moved to 3.11 and tried and use it, but it either fail to start or fail to stop or have a 10 seconds delay in it.

**Comment by janiversen** [comment]
You called asyncio.run to use ServerAsyncStop, that is not how asyncio works.

Pymodbus have been around for many years with many users, one of our benefits is the flexibility if you e.g. don’t like startServer, then you do not need to use it, use the server class instead. Since we stopped supporting python 3.5 and until today, pymodbus have had major changes just like python, so please accept that some things work differently than you are used to, for a lot of reasons.

**Reply by 28itay** [reply]
I know that somethings work differently and I'm okay with that, I just try and find a way to use it at it's current state, I tried to use `serve_forever` and `shutdown` like I used before in Python 3.5, but it didn't work for me, so I looked around to see how else I can use it, I tried using `StartTcpServer`, I tried using `StartAsyncTcpServer`, I tried using `ServerStop` and `ServerAsyncStop` in any combination.

The only combination that worked is using `ServerStop` with `StartTcpServer` but there is 10 seconds delay on `ServerStop`, so I reached out, I don't real care using async or not...

**Comment by janiversen** [comment]
Well ServerAsyncStop and StartAscyncTcpServer works as well, just see our examples.

And as I have now said multiple times, if you do not like ServerStop, copy the function and make your own, do not expect us to do that.
